rpcs3/util/expr.hpp

```diff
@@ -0,0 +1,298 @@
+	}
+
+	template <typename S0, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, SS&&... stack)
+	{
+		if (s0)
+		{
+			utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, std::forward<SS>(stack)...);
+		}
+		else
+		{
+			skip<S.data[I], S, I + 1>(std::forward<SS>(stack)...);
+		}
+	}
+};
+
+// Duplicate top stack value
+template <utf32_expr S, usz I>
+struct utf32_cmd<U':', S, I>
+{
+	template <typename S0, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, SS&&... stack)
+	{
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, std::forward<S0>(s0), std::forward<S0>(s0), std::forward<SS>(stack)...);
+	}
+};
+
+// Duplicate two top stack values
+template <utf32_expr S, usz I>
+struct utf32_cmd<U'#', S, I>
+{
+	template <typename S0, typename S1, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, S1&& s1, SS&&... stack)
+	{
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, std::forward<S0>(s0), std::forward<S0>(s1), std::forward<S0>(s0), std::forward<S1>(s1), std::forward<SS>(stack)...);
+	}
+};
+
+// Swap two top stack values
+template <utf32_expr S, usz I>
+struct utf32_cmd<U's', S, I>
+{
+	template <typename S0, typename S1, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, S1&& s1, SS&&... stack)
+	{
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, std::forward<S1>(s1), std::forward<S0>(s0), std::forward<SS>(stack)...);
+	}
+};
+
+// Rotate stack so the top becomes the bottom
+template <utf32_expr S, usz I>
+struct utf32_cmd<U',', S, I>
+{
+	template <typename S0, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, SS&&... stack)
+	{
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, std::forward<SS>(stack)..., std::forward<S0>(s0));
+	}
+};
+
+// Negate top stack value
+template <utf32_expr S, usz I>
+struct utf32_cmd<U'!', S, I>
+{
+	template <typename S0, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, SS&&... stack)
+	{
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, std::forward<S0>(s0) ? false : true, std::forward<SS>(stack)...);
+	}
+};
+
+#define REG_BIN_OP(ch, _func)\
+template <utf32_expr S, usz I>\
+struct utf32_cmd<ch, S, I>\
+{\
+	template <typename S0, typename S1, typename... SS>\
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, S1&& s1, SS&&... stack)\
+	{\
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, (_func)(std::forward<S0>(s0), std::forward<S1>(s1)), std::forward<SS>(stack)...);\
+	}\
+};\
+
+REG_BIN_OP(U'+', [](const auto& a, const auto& b){ return a + b; })
+REG_BIN_OP(U'-', [](const auto& a, const auto& b){ return a - b; })
+REG_BIN_OP(U'*', [](const auto& a, const auto& b){ return a * b; })
+REG_BIN_OP(U'/', [](const auto& a, const auto& b){ return a / b; })
+REG_BIN_OP(U'«', [](const auto& a, const auto& b){ return a << b; })
```

Comment:
Why these non-standard characters? In order to use this operator in the DSL, one would have to copy and paste it, making it tedious. It might be harder to parse multiple characters e.g. `>>` but it might be worth extending this PR to support operators as they are now, rather than these non-standard ones.

Reply:
Or just press Compose Key + <<, available on most systems and quite powerful.

Reply:
The question still remains, why though?
Developers are already used to the normal shift syntax, this just makes things more obscure...

Reply:
Literally where would that compose key be? Never in my life have I heard or seen such a thing.

Reply:
Theres no compose key on windows as far as I see, this function is unix specific.
Looks like you need extra tools for windows: https://superuser.com/questions/64178/compose-key-on-windows

Reply:
It won't be as obscure after it is properly documented besides making varying length operator tokens with c++ templates is even more insane. You'd need actual token parsing instead of simple per character transformation. Besides there already is token collision between `>` and `>>`, `<` and `<<` so those will have to changed or token delimiter would need to be added which will complicate things further than they already are.

Reply:
I can't comment on the implementation difficulty, I'm not really that experienced with C++.
Even if it is really well documented, it still is more obscure and harder to use.

Besides, is making devs on windows second class citizens (even more) really worth it?

Reply:
Not having compose key on windows out of the box is not really a sound argument since there are more than enough ways to add it within like 5 clicks with no further actions required. I personally prefer [wincompose](https://github.com/SamHocevar/wincompose).

Thou i agree that stack machine DSL seems a bit much for this. It is certainly fun but i doubt it would be more efficient than simply writing the lambda yourself. At the current point it seems more esoteric than practical, but i am not sure about the actual use case in rpcs3 codebase besides the one provided.

Reply:
The Windows equivalent is alt codes, but for those you have to memorize codes for each key.

Reply:
btw how about a new shiny github "discussion" thingy? should be useful here

Reply:
The main reason is that it's easy to parse one unicode character as one command. It's secondary to make it more compact.
Personally, I used to have configured Compose key on Caps Lock, but then I switched it with Menu key, with AltGr as Level3 shift and Right Win as Level5 (I don't use it though).